### ISSUE TITLE
Fork bomb prevention

### ISSUE THREAD

**Issue by yoshiokatsuneo** [docker/docker#6479]
I'm using docker with native driver. (It is much faster than lxc driver.)
Now, docker command have option "-m" to limit memory usage.
But, there is no way to prevent "fork bomb", I think.
Nowadays, linux kernel support "memory. kmem.limit_in_bytes" cgroups settings to prevent fork bomb.
It is quite nice if there is some way to set  "memory. kmem.limit_in_bytes" cgroups settings .
Or, probably it is even nice if there is generic way to specify cgroups settings.

One workaround is to use "ulimit -u", but it does not work well for root user or if there are multiple container with same user.

On ubuntu 14.04, we need to update kernel package to 3.13.0-27.50 or later to use "memory. kmem.limit_in_bytes" cgroups settings 
Ref: https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1309586

Thanks.

**Comment by jfrazelle** [comment]
ping @crosbymichael im going to label as black-belt hope thats cool

**Comment by nfritze** [comment]
Have been looking into this as well on our docker setup.  If we bump the pid/kernel-thread address space on the host a bit to give it breathing room:

echo $((2**22))> /proc/sys/kernel/pid_max
echo $((2**22))> /proc/sys/kernel/threads-max

then at that point it can grow enough that, if we limit the memory, it will be limited by that before it takes down the host.  For testing, was using docker 1.5, launching a container with:

docker run -t -i --rm --memory=200M --memory-swap=400M ubuntu:14.04 /bin/bash

and then triggering a forkbomb in it.  Inside, the container becomes unusable immediately.  Outside, on the host, it doesn't die (if we don't expand the pid/thread space, host does die here), but we're very slowly leaking memory.  Over several hours, it grew to using >2G of memory (virtual), and was still growing.  Not sure if there's some other setting I need to look into, for now, continuing to try to lock it down so it can't do this.

Any other suggestions welcome.

**Comment by ernetas** [comment]
@nfritze you could try working with ulimit to set reasonable limits for processes running within Docker. However, because you may be running multiple processes and some may need more, others may need less resources, and even worse, you cannot increase some of the limits within Docker after they have been decreased, this may or may not work for you. For me, what worked best was leaving everything as is, except for setting strict ulimit settings for processes that I was actually expecting to cause issues.

But this only works if you can set VERY strict ulimit settings. Otherwise, in my testing, both the host and the container were suffering from severe performance degradation (making it very difficult to strace/gdb (in the beginning, I didn't even knew which part of my code exactly was causing the fork bomb) or debug in any other way what's exactly happening) and sometimes even crashing, if the limits were too high.

**Comment by cpuguy83** [comment]
#9437 will allow you to set default ulimits for all containers as well as set custom ulimits per container.

**Comment by nfritze** [comment]
That sounds very promising, thank you!  Investigating.

**Comment by thaJeztah** [comment]
@cpuguy83 will this be solved now that https://github.com/docker/docker/pull/9437 has been merged?

**Comment by cpuguy83** [comment]
@yoshiokatsuneo does the new ulimit options from #9437 resolve this issue for you?

Setting ulimits should prevent fork bomb, but this is specifically asking for allowing to set kmem.limit_in_bytes

**Comment by frol** [comment]
@cpuguy83 ulimit sets limits for UID/GID, not per container. This option worked for me fine until I went from one container to dozens.

**Comment by cpuguy83** [comment]
There must be something in what we're doing with libcontainer to set this... my understanding is that ulimit is per process.

**Comment by cpuguy83** [comment]
> The nproc limit is a bit of a special case, in that it does sum over all the processes of a user. Nonetheless, it still applies per-process: when a process calls fork to create a new process, the call is denied if the number of processes belonging to the process's euid is would be larger than the process's RLIMIT_NPROC value.
http://unix.stackexchange.com/questions/55319/are-limits-conf-values-applied-on-a-per-process-basis

:( that's troublesome.

**Comment by frol** [comment]
@cpuguy83 It is. And kernel devs suggested to use `memory.kmem.limit_in_bytes` to prevent things like fork-bombs, but it doesn't seem to me like a very good approach.

There were 2 tries to implement nproc in cgroups, but both were rejected, but today I discovered the new version of this: https://lkml.org/lkml/2015/2/22/204

Relevant links:
* New patch which adds nproc cgroups subsystem: https://lkml.org/lkml/2015/2/22/204
* Old patch which tried to add task counter subsystem (was never accepted): https://lkml.org/lkml/2012/1/31/489
* Old response from Tejun Heo (kernel subsystem maintainer): https://lkml.org/lkml/2013/4/1/346

**Comment by frol** [comment]
Another piece of information is that `memory.kmem.limit_in_bytes` leads to a slab (unreclaimable) memory leakage. It is quite small (~2kb per group), but you will need to **reboot** your host to free up your memory.

For example, I have a project that uses `memory.kmem.limit_in_bytes` intensively (~50k new cgroups per hour) and it leads to 100mb leak per hour. That is crucial on a machine with 1gb of ram!

I have found only [one little note](http://www.linuxplumbersconf.org/2014/wp-content/uploads/2014/10/LPC2014_Containers.txt) on this case, that was made on "LPC 2014: Are containers that we have now secure enough?":

> - kmem accounting:
>    - kmem is everything except userspace memory.
>    - idea is to keep from hitting denial-of-service style resource exhaustion
>    - accounting already done as part of memory cgroup - memory.kmem.limit_in_bytes
>    - however, **lacks slab-shrinking support to reclaim memory** (targetting for Linux 3.19):
>        - won't account for page tables
>        - **fixable memory leaks still in patches**
>    - we end up with a container in ENOMEM loops

I've tested 4.1rc1 kernel and it still leaks, so I reached the guys who presented the talk, but haven't received any feedback yet.

**Comment by hqhq** [comment]
@frol kmem limit is not a good approach to resolve fork bomb, since it's not ready for product environment even on 3.18 (latest LTS), I don't know if we should add kmem limit support for Docker, if someone really need it, I can send a PR.
For your problem, you can wait for pid cgroup, this is the new version of task count cgroup, and most likely will be merge to upstream.
https://lwn.net/Articles/638651/

**Comment by frol** [comment]
@hqhq Yes, I know that it is not that I really wanted for preventing fork-bombs, but it is the only already working solution available in the upstream. `nproc` was my first attempt, but it failed as @cpuguy83 quoted, so I moved on `kmem.limit_in_bytes` and after a heavy load it failed also, but not as much as `nproc` though.

Nevertheless, I think `kmem.limit_in_bytes` would be a good option to have in Docker out-of-the-box as the leaks will be fixed eventually.

Here is the latest v10 patchset for pids subsystem: http://lkml.iu.edu/hypermail/linux/kernel/1504.2/01738.html
(I guess it will take at least two more iterations based on the thread till it gets to Linus, so I wouldn't expect it is merged into upstream before June :( ).

**Comment by frol** [comment]
**Update on `memory.kmem.limit_in_bytes`**

The leaks were actually fixed in 4.0 kernel with this commit: https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=d6e0b7fa11862433773d986b5f995ffdf47ce672

The reason I reported that they were not fixed was because there is still mislead in counting SUnreclaim. If I run my experiment with `memory.kmem.limit_in_bytes` and drop caches afterwards (`echo 2 > /proc/sys/vm/drop_caches`) the value of SUnreclaim shrinks to the value before the experiment! Thus, it doesn't leak!

@hqhq Taking in account that the leaks were fixed in the upstream, it can be safely added to Docker with a notice about kernel requirement (4.0+).

**Comment by hqhq** [comment]
@frol , then I think we can add kmem limit support for Docker, with kernel version 4.0+ as hard requirement. I'll do it in the following days.

**Comment by sunyc** [comment]
@vangie  We (Coding.net)  recently start working on preventing fork bomb on our docker based public web IDE service .  We will be able to share some info here.

**Comment by thaJeztah** [comment]
@sunyc awesome, greatly appreciated!

**Comment by frol** [comment]
@sunyc very promising. Thus, I will share my conclusions about fork-bomb prevention:

1. There is only one thing that is available out-of-the-box: `ulimit nproc`. While it sounds like it is a solution, it is not! ulimit nproc limits number of processes by UID in a host system, so if you run several containers with the same UID inside, your containers will interfere. Conclusion: `ulimit nproc` doesn't work.
1. The next step into this topic showed that there is a way to limit kernel memory usage via cgroup, which is directly involved on forking. Limit kmem (`memory.kmem.limit_in_bytes`) by 1MB preventes fork-bombs and works individually for each container (cgroup). Sounds good, right? However, there are some drawbacks:
  * Docker doesn't support limiting kmem via command line options, but @hqhq promised to do this (no out-of-the-box Docker support)
  * kmem includes not only info about forking, but a lot of different things like kernel-encryption, FS related stuff and others, so it is not a really good solution to prevent fork-bombs, as some other operations may require more than 1MB (kmem is only partial solution, doesn't work for everybody)
  * Kernels <4.0 has a kernel memory leakage when you set `memory.kmem.limit_in_bytes`, so you'd better be on the edge kernel, don't forget to compile AUFS... Here is the list of distribs with their status on the kernel+AUFS:
    * Ubuntu has mainline kernels without AUFS
    * Arch Linux has official kernel 4.0, but without AUFS (AUR has 4.0+AUFS, but it is painful to use kernel from AUR)
    * boot2docker latest release has kernel 3.18+AUFS, but they have already merged 4.0+AUFS into master, so we are waiting for the next release
   *  RancherOS potentially has all requirements right!
1. While `memory.kmem.limit_in_bytes` somewhat solved fork-bombs for me, it is not the ideal solution. There is another cutting-edge approach: [cgroup pid subsystem](http://lkml.iu.edu/hypermail/linux/kernel/1504.2/01738.html) (it is still work in progress, you can read some more information about this in the comments above). I haven't tried it at all, but it looks like exactly what we need here.

### My solution

I run docker containers with mounted `/sys/fs/cgroup` inside a container and use my fork of [`cgmemtime`](https://github.com/frol/cgmemtime) utility with `--memory-limit 64m --kernel-memory-limit 1m --kill-children-on-exit` options.

**Comment by frol** [comment]
"Great"... Docker 1.6.1 introduced security update which prohibit to mount `/sys` and as a result of it `/sys/fs/cgroup`, which I used for my solution... https://github.com/docker/docker/commit/2f54c1a352e23de04a4535d0fc1bd0cad66c7c15

**Comment by frol** [comment]
**Update on my previous reply about `/sys/fs/cgroup`**
The commit was reverted and my solution works on Docker 1.6.2!

**Comment by ChampionL** [comment]
@frol  ulimit set ulimits for every session ,not for uid http://linux.die.net/man/5/limits.conf,so I create two containers and this two containers exist in different session(they have different session-id),so I can set different ulimit for every container.I  set the nproc and open-files paramter,both methods  could prevent forkbomb. I am confused by your conclusion. so could you share how to conclude that? is there any mistake in my mind?

**Comment by frol** [comment]
@ChampionL Have you read the quote from @cpuguy83 ?

> > The nproc limit is a bit of a special case, in that it does sum over all the processes of a user. Nonetheless, it still applies per-process: when a process calls fork to create a new process, the call is denied if the number of processes belonging to the process's euid is would be larger than the process's RLIMIT_NPROC value.
> > http://unix.stackexchange.com/questions/55319/are-limits-conf-values-applied-on-a-per-process-basis

> :( that's troublesome.

Check out also this issue #12695. If you have a way to achieve proper limit setting by nproc, please provide commands.

**Comment by ChampionL** [comment]
thanks @frol ,ulimit could not prevent forkbomb, kmem prevent forkbomb .

**Comment by poelzi** [comment]
@frol have you tried starting the namespace first and executing the ulimit syscall inside the new namespace ?
(have not found out the flow from the patch, sorry ;))

**Comment by frol** [comment]
@poelzi what kind of namespace do you refer to? Just for your info, it is Docker daemon that sets up new containers and I cannot easily hack in the middle of this process to create any kind of namespace.

**Comment by jpconver** [comment]
Hi!
What about the solution in this post -> https://devlearnings.wordpress.com/2014/08/22/limiting-fork-bomb-in-docker/
If two containers with the same user are started they will interfere or the max process count is per container?
Thanks!

**Comment by yoshiokatsuneo** [comment]
> If two containers with the same user are started they will interfere or the max process count is per container?
nproc is not for per container, but for per user.

**Comment by cpuguy83** [comment]
Closing since #14006 is merged. You can now use `--kernel-memory` to prevent fork bombs.

**Comment by frol** [comment]
@cpuguy83 While `--kernel-memory` provides at least usable solution (in contrast to `ulimit nproc`), it is still only a half of the story. We cannot do anything with it just yet, but cgroup PIDs subsystem (CGROUP_PIDS kernel setting) is going to get merged to Linux Kernel 4.3. In my opinion, only when PIDs subsystem is there and Docker support for it is added, only then we can claim the fork-bomb prevention is covered.

**Comment by cpuguy83** [comment]
@frol fair enough, thanks!

**Comment by haches** [comment]
@frol @cpuguy83 Could you elaborate on how ```--kernel-memory``` can be used to prevent fork bombs? I couldn't find any good information on this. Will this be included in the next release? Thanks a lot.

**Comment by frol** [comment]
@haches I set kernel memory limit to 1MB (I haven't used `--kernel-memory` option yet since I have worked around this issue with `cgmemtime`). It might not work nice when your containered app uses kernel memory not only for processes, but also, for example, for kernel-level cryptography (I'm not very familiar with the topic of kernel memory, though).

**Comment by LK4D4** [comment]
We merged PID cgroup support to libcontainer, so I think now we can implement it's support in Docker as well.

**Comment by funkytaco** [comment]
nproc cgroup is supported in Docker 1.10, right? So this should be solved?

**Comment by cpuguy83** [comment]
Yep, thanks.

**Comment by yoshiokatsuneo** [comment]
@funkytaco @cpuguy83 Actually, not yet on Docker1.10, but it *will* come on 1.11.
  https://blog.docker.com/2016/02/docker-engine-1-10-security/

**Comment by cpuguy83** [comment]
@yoshiokatsuneo True, but it's already merged.

**Comment by jamespic** [comment]
@frol It's an understatement to say that it doesn't work nice with kernel-level cryptography. I tried running a fork bomb in docker with `--kernel-memory=4M` on a system with ecryptfs. When I ran it, it successfully defused, but my kernel remounted my root fs read-only, for reasons that weren't apparent at the time. When I rebooted, it dropped me into an initramfs shell, and wouldn't boot until I `fsck`ed the borked fs.
